README.md

```diff
@@ -269,6 +227,101 @@ Details:
+    }
+  ]
+});
+
+const result = await session.prompt("What is the weather in Seattle?");
+```
+
+In this example, the `tools` array defines a `getWeather` tool, specifying its name, description, input schema, and `execute` implementation. When the language model determines that a tool call is needed, the user agent invokes the `getWeather` tool's `execute()` function with the provided arguments and returns the result to the model, which can then incorporate it into its response.
+
+#### Tool return values
+
+The above example shows tools returning a string. (In fact, stringified JSON.) Models which support [multimodal inputs](#multimodal-inputs) might also support interpreting image or audio results from tool calls.
+
+Just like the `content` option to a `prompt()` call can accept either a string or an array of `{ type, value }` objects, web developer-provided tools can return either a string or such an array. Here's an example:
+
+```js
+let mutex, resolveMutex;
+
+const session = await LanguageModel.create({
+  tools: [
+    {
+      name: "grabKeyframe",
+      description: "Grab a keyframe from the video we're analyzing at the given time",
+      inputSchema: {
+        type: "number",
+        minimum: 0,
+        exclusiveMaximum: videoEl.duration
+      },
+      expectedOutputs: {
+        types: ["image"]
+      },
+      async execute(timestamp) {
+        if (mutex) {
+          // Since we're seeking a single video element, guard against concurrent calls.
+          await mutex;
+        }
+        try {
+          mutex = new Promise(r => resolveMutex = r);
+
+          if (Math.abs(videoEl.currentTime - timestamp) > 0.001) {
+            videoEl.currentTime = timestamp;
+            await new Promise(r => videoEl.addEventListener("seeked", r, { once: true }));
+          }
+          await new Promise(r => videoEl.requestVideoFrameCallback(r));
+
+          return [{ type: "image", value: videoEl }];
+        } finally {
+          resolveMutex();
+          mutex = null;
+        }
+      }
+    }
+  ]
+});
+```
+
+Note how the output types need to be specified in the tool definition, so that session creation can fail early if the model doesn't support processing multimodal tool outputs. If the return value contains non-text components without them being present in the tool specification, then the tool call will fail at prompting time, even if the model could support it.
+
+Similarly, expected output languages can be provided (via `expectedOutputs: { languages: ["ja" ] }`) or similar, to get an early failure if the model doesn't support processing tool outputs in those languages. However, unlike modalities, there is no prompt-time checking of the tool call result's languages.
+
+The above example shows a single-item array, but just like with prompt inputs, it's allowed to include multiple tool outputs. The same rules are followed as for inputs, e.g., concatenation of adjacent text chunks is done with a single space character.
```

Comment:
FWIW, line 258 below and chrome impl concatenates adjacent text chunks *without* a space character. I might not elude to a specific concatenation algorithm here. (but also, is something inconsistent to spec?)